File: backtest.js
const fs = require('fs');
const readline = require('readline');

async function realisticBacktest() {
    const data = [];
    const rl = readline.createInterface({
        input: fs.createReadStream('XAUUSDm_1m_2026-02-01_to_2026-04-01.csv'),
        crlfDelay: Infinity
    });

    let firstLine = true;
    for await (const line of rl) {
        if (firstLine) { firstLine = false; continue; }
        const parts = line.split(',');
        data.push({
            time: parts[0],
            open: parseFloat(parts[1]),
            high: parseFloat(parts[2]),
            low: parseFloat(parts[3]),
            close: parseFloat(parts[4]),
            volume: parseFloat(parts[5])
        });
    }

    let balance = 1000;
    const trades = [];
    let position = null;
    let entryPrice = 0;
    let sl = 0;
    let tp = 0;
    let entryTime = null;

    console.log("=== REALISTIC BACKTEST RESULTS ===");
    console.log(`Total candles loaded: ${data.length}`);
    console.log("Rules: Signal on close, Entry on next open, 0.01 slippage, No lookahead");
    console.log("=");

    for (let i = 3; i < data.length; i++) {
        const current = data[i];
        
        // Calculate RSI(5) - NO LOOKAHEAD, only uses past data
        const closes = data.slice(i-5, i+1).map(x => x.close);
        let gain = 0, loss = 0;
        for (let j = 1; j < closes.length; j++) {
            const delta = closes[j] - closes[j-1];
            if (delta > 0) gain += delta;
            else loss += Math.abs(delta);
        }
        gain /= 5;
        loss /= 5;
        const rs = loss === 0 ? 100 : gain / loss;
        const rsi = 100 - (100 / (1 + rs));

        // Previous 3 candle high/low
        const prev3 = data.slice(i-3, i);
        const prev3High = Math.max(...prev3.map(x => x.high));
        const prev3Low = Math.min(...prev3.map(x => x.low));

        // Entry logic - ONLY IF NO POSITION
        if (position === null && i + 1 < data.length) {
            const nextCandle = data[i+1];
            
            // BUY SIGNAL: Current candle CLOSED above prev 3 high + RSI < 30
            if (current.close > prev3High && rsi < 30) {
                entryPrice = nextCandle.open + 0.01; // +0.01 slippage (realistic)
                sl = entryPrice - 3.00;
                tp = entryPrice + 4.50;
                position = 'buy';
                entryTime = nextCandle.time;
                continue;
            }
            
            // SELL SIGNAL: Current candle CLOSED below prev 3 low + RSI > 70
            if (current.close < prev3Low && rsi > 70) {
                entryPrice = nextCandle.open - 0.01; // -0.01 slippage (realistic)
                sl = entryPrice + 3.00;
                tp = entryPrice - 4.50;
                position = 'sell';
                entryTime = nextCandle.time;
                continue;
            }
        }

        // Exit logic
        if (position !== null) {
            if (position === 'buy') {
                // Check SL first (always)
                if (current.low <= sl) {
                    const profit = (sl - entryPrice) * 200 * 0.02;
                    balance += profit;
                    trades.push({ type: 'buy', entry: entryPrice, exit: sl, profit, reason: 'SL', entryTime, exitTime: current.time });
                    position = null;
                }
                // Check TP
                else if (current.high >= tp) {
                    const profit = (tp - entryPrice) * 200 * 0.02;
                    balance += profit;
                    trades.push({ type: 'buy', entry: entryPrice, exit: tp, profit, reason: 'TP', entryTime, exitTime: current.time });
                    position = null;
                }
            }
            else if (position === 'sell') {
                // Check SL first
                if (current.high >= sl) {
                    const profit = (entryPrice - sl) * 200 * 0.02;
                    balance += profit;
                    trades.push({ type: 'sell', entry: entryPrice, exit: sl, profit, reason: 'SL', entryTime, exitTime: current.time });
                    position = null;
                }
                // Check TP
                else if (current.low <= tp) {
                    const profit = (entryPrice - tp) * 200 * 0.02;
                    balance += profit;
                    trades.push({ type: 'sell', entry: entryPrice, exit: tp, profit, reason: 'TP', entryTime, exitTime: current.time });
                    position = null;
                }
            }
        }
    }

    const wins = trades.filter(t => t.profit > 0);
    const losses = trades.filter(t => t.profit < 0);
    
    console.log(`Total trades: ${trades.length}`);
    console.log(`Winning trades: ${wins.length}`);
    console.log(`Losing trades: ${losses.length}`);
    console.log(`Win rate: ${(wins.length / trades.length * 100).toFixed(1)}%`);
    console.log(`Average win: $${(wins.reduce((a,b) => a + b.profit, 0) / wins.length).toFixed(2)}`);
    console.log(`Average loss: $${(losses.reduce((a,b) => a + b.profit, 0) / losses.length).toFixed(2)}`);
    console.log(`Net profit: $${(balance - 1000).toFixed(2)}`);
    console.log(`Average profit per day: $${((balance - 1000) / 60).toFixed(2)}`);
    
    const cumsum = trades.map((_, i) => trades.slice(0, i+1).reduce((a,b) => a + b.profit, 0));
    console.log(`Maximum drawdown: $${Math.min(...cumsum).toFixed(2)}`);
    
    console.log("\nFirst 10 trades (verify execution order):");
    trades.slice(0, 10).forEach((t, i) => {
        console.log(`${i+1}. ${t.type.toUpperCase()} | Entry: ${t.entry.toFixed(3)} | Exit: ${t.exit.toFixed(3)} | ${t.reason} | $${t.profit.toFixed(2)}`);
    });
}

realisticBacktest();

File: backtest.py
import pandas as pd
import numpy as np

# Read 1-minute data
df = pd.read_csv('XAUUSDm_1m_2026-02-01_to_2026-04-01.csv')
df['time'] = pd.to_datetime(df['time'])

# STRICT BACKTEST RULES:
# 1. Signal generated ONLY on candle CLOSE
# 2. Entry on NEXT candle OPEN
# 3. SL/TP checked on every subsequent candle
# 4. 0.1 pip slippage added (realistic Exness execution)
# 5. No repainting, no lookahead

def realistic_backtest(df):
    balance = 1000
    trades = []
    position = None
    entry_price = 0
    sl = 0
    tp = 0
    
    for i in range(3, len(df)):
        # Current candle is closed (i), we can calculate signals now
        current_close = df.iloc[i]['close']
        prev3_high = df.iloc[i-3:i]['high'].max()
        prev3_low = df.iloc[i-3:i]['low'].min()
        
        # RSI(5) calculation (no lookahead)
        delta = df.iloc[i-5:i+1]['close'].diff()
        gain = delta.where(delta > 0, 0).mean()
        loss = -delta.where(delta < 0, 0).mean()
        rs = gain / loss if loss != 0 else 100
        rsi = 100 - (100 / (1 + rs))
        
        # Entry signal - ONLY if no position
        if position is None:
            # Entry on NEXT CANDLE OPEN (i+1) if it exists
            if i + 1 >= len(df):
                continue
                
            # BUY SIGNAL: Close above prev 3 high + RSI < 30
            if current_close > prev3_high and rsi < 30:
                entry_price = df.iloc[i+1]['open'] + 0.01  # +0.01 slippage
                sl = entry_price - 3.00
                tp = entry_price + 4.50
                position = 'buy'
                entry_index = i + 1
                entry_time = df.iloc[i+1]['time']
                
            # SELL SIGNAL: Close below prev 3 low + RSI > 70
            elif current_close < prev3_low and rsi > 70:
                entry_price = df.iloc[i+1]['open'] - 0.01  # -0.01 slippage
                sl = entry_price + 3.00
                tp = entry_price - 4.50
                position = 'sell'
                entry_index = i + 1
                entry_time = df.iloc[i+1]['time']
        
        # Check exit conditions if in position
        else:
            current_high = df.iloc[i]['high']
            current_low = df.iloc[i]['low']
            current_close = df.iloc[i]['close']
            
            if position == 'buy':
                # Check stop loss first
                if current_low <= sl:
                    profit = (sl - entry_price) * 200 * 0.02  # 0.02 lots, XAU pip value = $200 per $1
                    balance += profit
                    trades.append({
                        'entry_time': entry_time,
                        'exit_time': df.iloc[i]['time'],
                        'type': 'buy',
                        'entry': entry_price,
                        'exit': sl,
                        'profit': profit,
                        'reason': 'SL'
                    })
                    position = None
                # Check take profit
                elif current_high >= tp:
                    profit = (tp - entry_price) * 200 * 0.02
                    balance += profit
                    trades.append({
                        'entry_time': entry_time,
                        'exit_time': df.iloc[i]['time'],
                        'type': 'buy',
                        'entry': entry_price,
                        'exit': tp,
                        'profit': profit,
                        'reason': 'TP'
                    })
                    position = None
                    
            elif position == 'sell':
                # Check stop loss first
                if current_high >= sl:
                    profit = (entry_price - sl) * 200 * 0.02
                    balance += profit
                    trades.append({
                        'entry_time': entry_time,
                        'exit_time': df.iloc[i]['time'],
                        'type': 'sell',
                        'entry': entry_price,
                        'exit': sl,
                        'profit': profit,
                        'reason': 'SL'
                    })
                    position = None
                # Check take profit
                elif current_low <= tp:
                    profit = (entry_price - tp) * 200 * 0.02
                    balance += profit
                    trades.append({
                        'entry_time': entry_time,
                        'exit_time': df.iloc[i]['time'],
                        'type': 'sell',
                        'entry': entry_price,
                        'exit': tp,
                        'profit': profit,
                        'reason': 'TP'
                    })
                    position = None
    
    return pd.DataFrame(trades), balance

trades_df, final_balance = realistic_backtest(df)

print("=== REALISTIC BACKTEST RESULTS ===")
print(f"Total trades: {len(trades_df)}")
print(f"Winning trades: {len(trades_df[trades_df['profit'] > 0])}")
print(f"Losing trades: {len(trades_df[trades_df['profit'] < 0])}")
print(f"Win rate: {len(trades_df[trades_df['profit'] > 0]) / len(trades_df) * 100:.1f}%")
print(f"Average win: ${trades_df[trades_df['profit'] > 0]['profit'].mean():.2f}")
print(f"Average loss: ${trades_df[trades_df['profit'] < 0]['profit'].mean():.2f}")
print(f"Net profit: ${final_balance - 1000:.2f}")
print(f"Average profit per day: ${(final_balance - 1000) / 60:.2f}")
print(f"Maximum drawdown: ${trades_df['profit'].cumsum().min():.2f}")
print("\nFirst 10 trades to verify execution:")
print(trades_df.head(10))